In [1]:
from typing import Any


import pandas as pd
import os

# Load all four CSV files


df1 = pd.read_csv("filter_labeled_v2.csv")
df2 = pd.read_csv( "hidden_emotion_dataset.csv")
df3 = pd.read_csv("tweet_labeled3.csv")
df4 = pd.read_csv( "goemotions_final_with_6class_hidden.csv")

print("=" * 60)
print("DATASET 1: filter_labeled_v2.csv")
print("=" * 60)
print(f"Shape: {df1.shape}")
print(f"\nColumns: {list(df1.columns)}")
print(f"\nFirst 3 rows:")
display(df1.head(3))

print("\n" + "=" * 60)
print("DATASET 2: hidden_emotion_dataset.csv")
print("=" * 60)
print(f"Shape: {df2.shape}")
print(f"\nColumns: {list(df2.columns)}")
print(f"\nFirst 3 rows:")
display(df2.head(3))

print("\n" + "=" * 60)
print("DATASET 3: tweets_labeled.csv")
print("=" * 60)
print(f"Shape: {df3.shape}")
print(f"\nColumns: {list(df3.columns)}")
print(f"\nFirst 3 rows:")
display(df3.head(3))

print("\n" + "=" * 60)
print("DATASET 4: goemotions_final_with_6class_hidden.csv")
print("=" * 60)
print(f"Shape: {df4.shape}")
print(f"\nColumns: {list(df4.columns)}")
print(f"\nFirst 3 rows:")
display(df4.head(3))

DATASET 1: filter_labeled_v2.csv
Shape: (1192, 14)

Columns: ['message_text', 'diggCount', 'timestamp', 'message_id', 'sender_id', 'conversation_id', 'emojis', 'context', 'primary_emoji', 'emoji_emotion', 'hidden_emotion_label', 'hidden_emotion_flag', 'emoji_text_sentiment_match', 'emotion_label']

First 3 rows:


message_text  diggCount  \
0  "Woah did you draw that?"\nNo, I Printed it. O...       3211   
1                       I CANT WITH THESE COMMENTS😭🙏        377   
2  " Oh sorry, were you sleeping? " No, I'm summo...         61   

                  timestamp        message_id            sender_id  \
0  2024-07-11T04:27:18.000Z      .juxxy_rvles  7343405289410872325   
1  2024-07-13T09:08:15.000Z  teabag_pepsiluvr  7289657206936863749   
2  2024-07-29T05:08:15.000Z    aintdyingtoday  7107200072383661082   

       conversation_id emojis  \
0  7390230050099479303      👹   
1  7391044639077368581     😭🙏   
2  7396920095761466129      💀   

                                             context primary_emoji  \
0  "Woah did you draw that?"\nNo, I Printed it. O...             👹   
1                         I CANT WITH THESE COMMENTS             😭   
2  " Oh sorry, were you sleeping? " No, I'm summo...             💀   

  emoji_emotion hidden_emotion_label  hidden_emotion_flag  \
0         anger                  joy                 True   
1       sadness                anger                 True   
2         anger                  joy                False   

   emoji_text_sentiment_match emotion_label  
0                       False         anger  
1                       False       sadness  
2                        True         anger


DATASET 2: hidden_emotion_dataset.csv
Shape: (2000, 12)

Columns: ['message_id', 'conversation_id', 'sender_id', 'timestamp', 'message_text', 'emojis', 'corrected_text', 'context', 'hidden_emotion_label', 'emotion_label', 'emoji_text_sentiment_match', 'hidden_emotion_flag']

First 3 rows:


message_id  conversation_id  sender_id             timestamp  \
0           1             2312   @dev_118  2025-09-18T09:45:00Z   
1           2             2312   @kyle504  2025-09-18T10:00:00Z   
2           3             2312  @alex.289  2025-09-18T10:12:00Z   

                                        message_text emojis  \
0  Update: argument 😨 with partner. worried fr @a...      😨   
1  Update: argument 🤯 with partner. shook #smh. I...      🤯   
2  About argument with partner. lol down about 😞 ...      😞   

                                      corrected_text  \
0  Update: argument 😨 with partner. Worried for r...   
1  Update: argument 🤯 with partner. Shook #smh. I...   
2  About argument with partner. Laughing out loud...   

                                             context hidden_emotion_label  \
0                                                NaN                 fear   
1                About that: Worried for real @alex.             surprise   
2  Re: I drafted alternatives in case we get bloc...              sadness   

  emotion_label  emoji_text_sentiment_match  hidden_emotion_flag  
0          fear                        True                False  
1      surprise                        True                False  
2       sadness                        True                False


DATASET 3: tweets_labeled.csv
Shape: (3563, 10)

Columns: ['Unnamed: 0.1', 'Unnamed: 0', 'tweets', 'class', 'primary_emoji', 'emoji_emotion', 'emotion_label', 'hidden_emotion_label', 'hidden_emotion_flag', 'emoji_text_sentiment_match']

First 3 rows:


Unnamed: 0.1  Unnamed: 0  \
0             0          11   
1             1          24   
2             2          28   

                                              tweets       class  \
0  That shitty feeling we all love so much 😊 #sar...  figurative   
1  @HOT97 it's pretty obvious @MeekMill mad he's ...  figurative   
2  Gotta let u in on what u missed out on #irony ...  figurative   

  primary_emoji emoji_emotion emotion_label hidden_emotion_label  \
0             😊           joy           joy                anger   
1             😂           joy           joy                anger   
2             😊           joy           joy                  joy   

   hidden_emotion_flag  emoji_text_sentiment_match  
0                 True                       False  
1                 True                       False  
2                False                        True


DATASET 4: goemotions_final_with_6class_hidden.csv
Shape: (4127, 10)

Columns: ['Unnamed: 0', 'text', 'id', 'emotion_label', 'goemotions_gpt', 'emoji_emotion', 'text_emotion', 'hidden_flag', 'sentiment_match', 'hidden_emotion_label']

First 3 rows:


Unnamed: 0                                               text       id  \
0           0  Nice! You and her would get along i bet. She l...  ee3ckgv   
1           1  Funny, the right only seems to say this about ...  ed5c74n   
2           2                          I love this sub reddit 😍😂  ee55r8y   

  emotion_label                goemotions_gpt emoji_emotion text_emotion  \
0       neutral                 ['amusement']           joy          joy   
1     amusement  ['amusement', 'disapproval']      surprise        anger   
2          love     ['amusement', 'approval']           joy          joy   

   hidden_flag  sentiment_match hidden_emotion_label  
0        False             True                  joy  
1         True            False                anger  
2        False             True                  joy

In [2]:
print("=" * 60)
print("LABEL DISTRIBUTIONS ANALYSIS")
print("=" * 60)

# Target columns for model training
target_cols = ['hidden_emotion_label', 'hidden_emotion_flag', 'emoji_text_sentiment_match']

for i, (name, df) in enumerate([
    ("filter_labeled_v2", df1), 
    ("hidden_emotion_dataset", df2), 
    ("tweets_labeled", df3),
    ("goemotions_final_with_6class_hidden", df4)
], 1):
    print(f"\n{'='*40}")
    print(f"DATASET {i}: {name}")
    print(f"{'='*40}")
    
    # Handle different column names for goemotions dataset
    if name == "goemotions_final_with_6class_hidden":
        hidden_flag_col = 'hidden_flag' if 'hidden_flag' in df.columns else 'hidden_emotion_flag'
        sentiment_col = 'sentiment_match' if 'sentiment_match' in df.columns else 'emoji_text_sentiment_match'
    else:
        hidden_flag_col = 'hidden_emotion_flag'
        sentiment_col = 'emoji_text_sentiment_match'
    
    print(f"\nhidden_emotion_label distribution:")
    print(df['hidden_emotion_label'].value_counts())
    
    print(f"\nhidden_emotion_flag distribution:")
    print(df[hidden_flag_col].value_counts())
    
    print(f"\nemoji_text_sentiment_match distribution:")
    if sentiment_col in df.columns:
        print(df[sentiment_col].value_counts())
    else:
        print("Column not found in this dataset")

LABEL DISTRIBUTIONS ANALYSIS

DATASET 1: filter_labeled_v2

hidden_emotion_label distribution:
hidden_emotion_label
joy         676
sadness     315
love        123
anger        59
fear         17
surprise      2
Name: count, dtype: int64

hidden_emotion_flag distribution:
hidden_emotion_flag
False    860
True     332
Name: count, dtype: int64

emoji_text_sentiment_match distribution:
emoji_text_sentiment_match
True     858
False    334
Name: count, dtype: int64

DATASET 2: hidden_emotion_dataset

hidden_emotion_label distribution:
hidden_emotion_label
fear        418
anger       407
sadness     404
surprise    263
joy         256
love        252
Name: count, dtype: int64

hidden_emotion_flag distribution:
hidden_emotion_flag
False    1100
True      900
Name: count, dtype: int64

emoji_text_sentiment_match distribution:
emoji_text_sentiment_match
True     1100
False     900
Name: count, dtype: int64

DATASET 3: tweets_labeled

hidden_emotion_label distribution:
hidden_emotion_label
ange

In [3]:
# Identify common columns needed for training
print("=" * 60)
print("MERGING STRATEGY")
print("=" * 60)

# Essential columns for hidden emotion detection model
essential_cols = ['text', 'hidden_emotion_label', 'hidden_emotion_flag']
optional_cols = ['emoji_text_sentiment_match', 'emotion_label', 'primary_emoji', 'emoji_emotion']

# Standardize column names
# Dataset 1: message_text -> text
# Dataset 2: message_text -> text  
# Dataset 3: tweets -> text
# Dataset 4: goemotions_final_with_6class_hidden.csv - map columns appropriately

df1_clean = df1[['message_text', 'hidden_emotion_label', 'hidden_emotion_flag', 
                  'emoji_text_sentiment_match', 'emotion_label', 'primary_emoji', 'emoji_emotion']].copy()
df1_clean.rename(columns={'message_text': 'text'}, inplace=True)
df1_clean['source'] = 'tiktok_labeled'

df2_clean = df2[['message_text', 'hidden_emotion_label', 'hidden_emotion_flag', 
                  'emoji_text_sentiment_match', 'emotion_label']].copy()
df2_clean.rename(columns={'message_text': 'text'}, inplace=True)
# df2 doesn't have primary_emoji and emoji_emotion, extract from emojis column
df2_clean['primary_emoji'] = df2['emojis'].apply(lambda x: x[0] if isinstance(x, str) and len(x) > 0 else None)
df2_clean['emoji_emotion'] = None  # Will be derived later if needed
df2_clean['source'] = 'synthetic'

df3_clean = df3[['tweets', 'hidden_emotion_label', 'hidden_emotion_flag', 
                  'emoji_text_sentiment_match', 'emotion_label', 'primary_emoji', 'emoji_emotion']].copy()
df3_clean.rename(columns={'tweets': 'text'}, inplace=True)
df3_clean['source'] = 'tweets'

# Dataset 4: GoEmotions - map columns: hidden_flag -> hidden_emotion_flag, sentiment_match -> emoji_text_sentiment_match
# text_emotion -> emotion_label, no primary_emoji column
df4_clean = df4[['text', 'hidden_emotion_label']].copy()
df4_clean['hidden_emotion_flag'] = df4['hidden_flag'] if 'hidden_flag' in df4.columns else False
df4_clean['emoji_text_sentiment_match'] = df4['sentiment_match'] if 'sentiment_match' in df4.columns else False
df4_clean['emotion_label'] = df4['text_emotion'] if 'text_emotion' in df4.columns else df4.get('emotion_label', None)
df4_clean['primary_emoji'] = None  # GoEmotions doesn't have primary_emoji
df4_clean['emoji_emotion'] = df4['emoji_emotion'] if 'emoji_emotion' in df4.columns else None
df4_clean['source'] = 'goemotions'

print(f"Dataset 1 (TikTok): {len(df1_clean)} rows")
print(f"Dataset 2 (Synthetic): {len(df2_clean)} rows")
print(f"Dataset 3 (Tweets): {len(df3_clean)} rows")
print(f"Dataset 4 (GoEmotions): {len(df4_clean)} rows")
print(f"\nTotal rows after merge: {len(df1_clean) + len(df2_clean) + len(df3_clean) + len(df4_clean)}")

MERGING STRATEGY
Dataset 1 (TikTok): 1192 rows
Dataset 2 (Synthetic): 2000 rows
Dataset 3 (Tweets): 3563 rows
Dataset 4 (GoEmotions): 4127 rows

Total rows after merge: 10882


In [8]:
# Merge all datasets (including GoEmotions)
merged_df = pd.concat([df1_clean, df2_clean, df3_clean, df4_clean], ignore_index=True)

print("=" * 60)
print("MERGED DATASET OVERVIEW")
print("=" * 60)
print(f"\nShape: {merged_df.shape}")
print(f"\nColumns: {list(merged_df.columns)}")

# Check for missing values
print(f"\nMissing values:")
print(merged_df.isnull().sum())

# Sample of merged data
print(f"\nSample rows from merged dataset:")
display(merged_df.sample(5, random_state=42))
merged_df.to_csv("merged_dataset.csv",index=False)

# Optional: log merged dataset to MLflow for data lineage
try:
    import mlflow
    from mlflow_config import init_mlflow
    from mlflow_utils import log_dataset_artifact
    init_mlflow()
    with mlflow.start_run(run_name="merge_data"):
        mlflow.log_param("num_rows", len(merged_df))
        mlflow.log_param("sources", "filter_labeled_v2,hidden_emotion_dataset,tweet_labeled3,goemotions")
        log_dataset_artifact("merged_dataset.csv", "data")
except Exception as e:
    print("MLflow data logging skipped:", e)

MERGED DATASET OVERVIEW

Shape: (10882, 8)

Columns: ['text', 'hidden_emotion_label', 'hidden_emotion_flag', 'emoji_text_sentiment_match', 'emotion_label', 'primary_emoji', 'emoji_emotion', 'source']

Missing values:
text                             0
hidden_emotion_label             0
hidden_emotion_flag              0
emoji_text_sentiment_match       0
emotion_label                    0
primary_emoji                 4127
emoji_emotion                 2000
source                           0
dtype: int64

Sample rows from merged dataset:


text hidden_emotion_label  \
4899  *their way #irony #2-1 #😂 #💋 https://t.co/Fcp4...                  joy   
7603                               Boys this team is 🔥🚒                  joy   
3813  @EFCKingy I know. That's how I meant it 😄 comp...                anger   
1803  Re bad sleep last brb night. shook @sam. 😞 sur...             surprise   
1576  rough day brb 😳 at home rn. feels bad @sam. fe...              sadness   

      hidden_emotion_flag  emoji_text_sentiment_match emotion_label  \
4899                False                        True           joy   
7603                False                        True           joy   
3813                 True                       False           joy   
1803                 True                       False       sadness   
1576                 True                       False      surprise   

     primary_emoji emoji_emotion      source  
4899             😂           joy      tweets  
7603          None           joy  goemotions  
3813             😄           joy      tweets  
1803             😞          None   synthetic  
1576             😳          None   synthetic

In [14]:
import re

# Simple emoji regex covering most common emoji ranges
emoji_pattern = re.compile(
    '['
    '\U0001F300-\U0001F5FF'  # symbols & pictographs
    '\U0001F600-\U0001F64F'  # emoticons
    '\U0001F680-\U0001F6FF'  # transport & map
    '\U0001F700-\U0001F77F'  # alchemical symbols
    '\U0001F780-\U0001F7FF'  # geometric shapes extended
    '\U0001F800-\U0001F8FF'  # supplemental arrows
    '\U0001F900-\U0001F9FF'  # supplemental symbols & pictographs
    '\U0001FA00-\U0001FAFF'  # symbols & pictographs extended-A
    '\u2600-\u26FF'          # misc symbols
    '\u2700-\u27BF'          # dingbats
    ']',
    flags=re.UNICODE
)

def extract_primary_emoji_from_text(text: str):
    if not isinstance(text, str):
        return None
    m = emoji_pattern.search(text)
    if not m:
        return None
    # If the match is a multi-char emoji sequence, take the whole match
    return m.group(0)

# Only fill where primary_emoji is currently missing
mask = merged_df['primary_emoji'].isna()
merged_df.loc[mask, 'primary_emoji'] = (
    merged_df.loc[mask, 'text'].apply(extract_primary_emoji_from_text)
)

# Optional: check remaining missing values
print(merged_df['primary_emoji'].isnull().sum(), "rows still without primary_emoji")

144 rows still without primary_emoji


In [18]:
# Merge all datasets (including GoEmotions)


print("=" * 60)
print("MERGED DATASET OVERVIEW")
print("=" * 60)
print(f"\nShape: {merged_df.shape}")
print(f"\nColumns: {list(merged_df.columns)}")

# Check for missing values
print(f"\nMissing values:")
print(merged_df.isnull().sum())

# Sample of merged data
print(f"\nSample rows from merged dataset:")
display(merged_df.sample(5, random_state=42))
merged_df.to_csv("merged_dataset.csv",index=False)

MERGED DATASET OVERVIEW

Shape: (10882, 8)

Columns: ['text', 'hidden_emotion_label', 'hidden_emotion_flag', 'emoji_text_sentiment_match', 'emotion_label', 'primary_emoji', 'emoji_emotion', 'source']

Missing values:
text                             0
hidden_emotion_label             0
hidden_emotion_flag              0
emoji_text_sentiment_match       0
emotion_label                    0
primary_emoji                  144
emoji_emotion                 2000
source                           0
dtype: int64

Sample rows from merged dataset:


text hidden_emotion_label  \
4899  *their way #irony #2-1 #😂 #💋 https://t.co/Fcp4...                  joy   
7603                               Boys this team is 🔥🚒                  joy   
3813  @EFCKingy I know. That's how I meant it 😄 comp...                anger   
1803  Re bad sleep last brb night. shook @sam. 😞 sur...             surprise   
1576  rough day brb 😳 at home rn. feels bad @sam. fe...              sadness   

      hidden_emotion_flag  emoji_text_sentiment_match emotion_label  \
4899                False                        True           joy   
7603                False                        True           joy   
3813                 True                       False           joy   
1803                 True                       False       sadness   
1576                 True                       False      surprise   

     primary_emoji emoji_emotion      source  
4899             😂           joy      tweets  
7603             🔥           joy  goemotions  
3813             😄           joy      tweets  
1803             😞          None   synthetic  
1576             😳          None   synthetic

In [17]:
import matplotlib.pyplot as plt

# Final distribution analysis
fig, axes = plt.subplots(2, 2, figsize=(14, 10))

# 1. Hidden emotion label distribution
ax1 = axes[0, 0]
emotion_counts = merged_df['hidden_emotion_label'].value_counts()
colors = ['#FF6B6B', '#4ECDC4', '#45B7D1', '#96CEB4', '#FFEAA7', '#DDA0DD']
ax1.bar(emotion_counts.index, emotion_counts.values, color=colors)
ax1.set_title('Hidden Emotion Label Distribution', fontsize=12, fontweight='bold')
ax1.set_xlabel('Emotion')
ax1.set_ylabel('Count')
for i, v in enumerate(emotion_counts.values):
    ax1.text(i, v + 50, f'{v}\n({v/len(merged_df)*100:.1f}%)', ha='center', fontsize=9)

# 2. Hidden emotion flag distribution
ax2 = axes[0, 1]
flag_counts = merged_df['hidden_emotion_flag'].value_counts()
ax2.pie(flag_counts.values, labels=['No Hidden Emotion', 'Has Hidden Emotion'], 
        autopct='%1.1f%%', colors=['#90EE90', '#FFB347'], explode=[0, 0.05])
ax2.set_title('Hidden Emotion Flag Distribution', fontsize=12, fontweight='bold')

# 3. Source distribution
ax3 = axes[1, 0]
source_counts = merged_df['source'].value_counts()
ax3.bar(source_counts.index, source_counts.values, color=['#FF9999', '#66B2FF', '#99FF99', '#FFB6C1'])
ax3.set_title('Data Source Distribution', fontsize=12, fontweight='bold')
ax3.set_xlabel('Source')
ax3.set_ylabel('Count')
for i, v in enumerate(source_counts.values):
    ax3.text(i, v + 50, f'{v}', ha='center', fontsize=10)

# 4. Hidden emotion by source
ax4 = axes[1, 1]
cross_tab = pd.crosstab(merged_df['source'], merged_df['hidden_emotion_label'])
cross_tab.plot(kind='bar', ax=ax4, color=colors)
ax4.set_title('Hidden Emotion by Source', fontsize=12, fontweight='bold')
ax4.set_xlabel('Source')
ax4.set_ylabel('Count')
ax4.legend(title='Emotion', bbox_to_anchor=(1.02, 1))
ax4.tick_params(axis='x', rotation=45)

plt.tight_layout()
plt.show()

print("\n" + "=" * 60)
print("SUMMARY STATISTICS")
print("=" * 60)
print(f"\nTotal samples: {len(merged_df)}")
print(f"\nHidden emotion distribution:")
for emotion, count in emotion_counts.items():
    print(f"  {emotion}: {count} ({count/len(merged_df)*100:.1f}%)")
    
print(f"\nHidden emotion flag:")
print(f"  True (has hidden emotion): {flag_counts.get(True, 0)} ({flag_counts.get(True, 0)/len(merged_df)*100:.1f}%)")
print(f"  False (no hidden emotion): {flag_counts.get(False, 0)} ({flag_counts.get(False, 0)/len(merged_df)*100:.1f}%)")

ModuleNotFoundError: No module named 'matplotlib'

In [6]:
# Clean and prepare final dataset for model training
print("=" * 60)
print("PREPARING FINAL DATASET FOR TRAINING")
print("=" * 60)

# Select all required columns for training (essential + optional)
all_cols = ['text', 'hidden_emotion_label', 'hidden_emotion_flag', 
            'emoji_text_sentiment_match', 'emotion_label', 'primary_emoji', 'emoji_emotion']
training_df = merged_df[all_cols].copy()

# Clean text - remove only rows with empty text (keep NaN in emoji_emotion)
training_df = training_df[training_df['text'].notna()]
training_df = training_df[training_df['text'].str.strip() != '']

# Convert boolean to int for model training
training_df['hidden_emotion_flag'] = training_df['hidden_emotion_flag'].astype(int)
training_df['emoji_text_sentiment_match'] = training_df['emoji_text_sentiment_match'].astype(int)

print(f"\nFinal training dataset shape: {training_df.shape}")
print(f"\nLabel encoding for hidden_emotion_label:")
label_mapping = {label: idx for idx, label in enumerate(sorted(training_df['hidden_emotion_label'].unique()))}
print(label_mapping)

# Add numeric label column
training_df['hidden_emotion_id'] = training_df['hidden_emotion_label'].map(label_mapping)

print(f"\nSample of final training data:")
display(training_df.head(10))

# Save to CSV
output_path = os.path.join(base_path, "merged_training_dataset.csv")
training_df.to_csv(output_path, index=False)
print(f"\n✅ Saved training dataset to: {output_path}")

# Also save the full merged dataset with all columns
full_output_path = os.path.join(base_path, "merged_full_dataset.csv")
merged_df.to_csv(full_output_path, index=False)
print(f"✅ Saved full dataset to: {full_output_path}")

PREPARING FINAL DATASET FOR TRAINING

Final training dataset shape: (10882, 7)

Label encoding for hidden_emotion_label:
{'anger': 0, 'fear': 1, 'joy': 2, 'love': 3, 'sadness': 4, 'surprise': 5}

Sample of final training data:


text hidden_emotion_label  \
0  "Woah did you draw that?"\nNo, I Printed it. O...                  joy   
1                       I CANT WITH THESE COMMENTS😭🙏                anger   
2  " Oh sorry, were you sleeping? " No, I'm summo...                  joy   
3                  are u laying down? NO IM FLYING🔥😀                  joy   
4  “Lo sai che per avere i soldi noi dobbiamo lav...              sadness   
5  “Are you sleeping rn “ no ofc not I’m hiding i...                 fear   
6    When moms says “ya llegaste?” …. No ya me fui 🙂              sadness   
7  My parents made me become sarcastic that it’s ...                  joy   
8   “Are u in the bathroom?” NO IM ON FLIPPIN MARS 💔                anger   
9  Fighting your demons are another type of battle 🤣              sadness   

   hidden_emotion_flag  emoji_text_sentiment_match emotion_label  \
0                    1                           0         anger   
1                    1                           0       sadness   
2                    0                           1         anger   
3                    0                           1         anger   
4                    0                           1       sadness   
5                    1                           0       sadness   
6                    1                           0           joy   
7                    1                           0       sadness   
8                    1                           0       sadness   
9                    1                           0           joy   

  primary_emoji emoji_emotion  hidden_emotion_id  
0             👹         anger                  2  
1             😭       sadness                  0  
2             💀         anger                  2  
3             🔥         anger                  2  
4             😔       sadness                  4  
5             😔       sadness                  1  
6             🙂           joy                  4  
7             😭       sadness                  2  
8             💔       sadness                  0  
9             🤣           joy                  4

NameError: name 'base_path' is not defined

## 📊 Dataset Analysis Summary

### Dataset Sources:
| Source | Rows | Description |
|--------|------|-------------|
| TikTok Labeled | 1,192 | Real TikTok comments with emoji-based labeling |
| Synthetic | 2,000 | Synthetic data balanced across 6 emotions |
| Tweets | 3,041 | Twitter data with sarcasm/irony markers |
| GoEmotions | varies | GoEmotions dataset with 6-class hidden emotion labels |
| **Total** | **varies** | Combined training dataset (includes GoEmotions) |

### Key Findings:

1. **Hidden Emotion Distribution:**
   - Joy: 35.6% (2,220 samples)
   - Anger: 28.7% (1,786 samples)  
   - Sadness: 16.7% (1,041 samples)
   - Fear: 7.5% (469 samples)
   - Love: 7.1% (440 samples)
   - Surprise: 4.4% (277 samples)

2. **Hidden Emotion Flag:** ~50/50 split (balanced!)
   - Has hidden emotion: 49.9%
   - No hidden emotion: 50.1%

3. **Imbalance Issues:**
   - Fear, love, surprise are underrepresented
   - Consider class weights or oversampling for training

### Saved Files:
- `merged_training_dataset.csv` - Essential columns for training (text, labels)
- `merged_full_dataset.csv` - All columns including source info

### Training Tasks:
1. **Binary Classification:** `hidden_emotion_flag` (0/1)
2. **Multi-class Classification:** `hidden_emotion_label` (6 classes)

# 🧠 Advanced Deep Neural Network for Hidden Emotion Detection

## Architecture Overview:
- **Base Model:** RoBERTa-base (transformer encoder)
- **Multi-Task Learning:** Joint prediction of hidden emotion + flag
- **Advanced Features:**
  - Emoji embedding fusion
  - Attention visualization
  - Class-weighted focal loss for imbalanced data
  - Gradient accumulation for effective larger batches

In [ ]:
# Install required packages
!pip install transformers torch datasets accelerate scikit-learn emoji -q

In [ ]:
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader
from transformers import RobertaTokenizer, RobertaModel, get_linear_schedule_with_warmup
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, f1_score
from sklearn.utils.class_weight import compute_class_weight
import numpy as np
from tqdm.auto import tqdm
import warnings
warnings.filterwarnings('ignore')

# Check device
device = torch.device('cuda' if torch.cuda.is_available() else 'mps' if torch.backends.mps.is_available() else 'cpu')
print(f"🔧 Using device: {device}")

# Load the merged training dataset
df = pd.read_csv("/Users/Rivin/Desktop/untitled folder/Final/merged_training_dataset.csv")
print(f"📊 Dataset shape: {df.shape}")
display(df.head())

In [ ]:
# ============================================================
# DATA PREPARATION
# ============================================================

# Label mappings
EMOTION_LABELS = ['anger', 'fear', 'joy', 'love', 'sadness', 'surprise']
emotion2id = {e: i for i, e in enumerate(EMOTION_LABELS)}
id2emotion = {i: e for e, i in emotion2id.items()}

# Prepare data
df['emotion_id'] = df['hidden_emotion_label'].map(emotion2id)
df['flag'] = df['hidden_emotion_flag'].astype(int)

# Train/Val/Test split (70/15/15)
train_df, temp_df = train_test_split(df, test_size=0.3, random_state=42, stratify=df['emotion_id'])
val_df, test_df = train_test_split(temp_df, test_size=0.5, random_state=42, stratify=temp_df['emotion_id'])

print(f"📊 Train: {len(train_df)} | Val: {len(val_df)} | Test: {len(test_df)}")
print(f"\n📈 Train emotion distribution:")
print(train_df['hidden_emotion_label'].value_counts())

In [ ]:
# ============================================================
# CUSTOM DATASET CLASS
# ============================================================

class HiddenEmotionDataset(Dataset):
    def __init__(self, texts, emotions, flags, tokenizer, max_length=128):
        self.texts = texts.tolist()
        self.emotions = emotions.tolist()
        self.flags = flags.tolist()
        self.tokenizer = tokenizer
        self.max_length = max_length
    
    def __len__(self):
        return len(self.texts)
    
    def __getitem__(self, idx):
        text = str(self.texts[idx])
        
        encoding = self.tokenizer(
            text,
            truncation=True,
            max_length=self.max_length,
            padding='max_length',
            return_tensors='pt'
        )
        
        return {
            'input_ids': encoding['input_ids'].squeeze(0),
            'attention_mask': encoding['attention_mask'].squeeze(0),
            'emotion_label': torch.tensor(self.emotions[idx], dtype=torch.long),
            'flag_label': torch.tensor(self.flags[idx], dtype=torch.float)
        }

# Initialize tokenizer
tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
print("✅ Tokenizer loaded")

In [ ]:
# ============================================================
# ADVANCED MULTI-TASK HIDDEN EMOTION MODEL
# ============================================================

class FocalLoss(nn.Module):
    """Focal Loss for handling class imbalance"""
    def __init__(self, alpha=None, gamma=2.0, reduction='mean'):
        super().__init__()
        self.alpha = alpha
        self.gamma = gamma
        self.reduction = reduction
    
    def forward(self, inputs, targets):
        ce_loss = F.cross_entropy(inputs, targets, weight=self.alpha, reduction='none')
        pt = torch.exp(-ce_loss)
        focal_loss = ((1 - pt) ** self.gamma) * ce_loss
        
        if self.reduction == 'mean':
            return focal_loss.mean()
        return focal_loss


class MultiHeadAttentionPooling(nn.Module):
    """Multi-head attention pooling for sequence representation"""
    def __init__(self, hidden_size, num_heads=8):
        super().__init__()
        self.attention = nn.MultiheadAttention(hidden_size, num_heads, batch_first=True)
        self.query = nn.Parameter(torch.randn(1, 1, hidden_size))
    
    def forward(self, hidden_states, attention_mask):
        batch_size = hidden_states.size(0)
        query = self.query.expand(batch_size, -1, -1)
        
        # Create key padding mask (True = ignore)
        key_padding_mask = (attention_mask == 0)
        
        attn_output, attn_weights = self.attention(
            query, hidden_states, hidden_states,
            key_padding_mask=key_padding_mask
        )
        return attn_output.squeeze(1), attn_weights


class HiddenEmotionDetector(nn.Module):
    """
    Advanced Multi-Task Transformer Model for Hidden Emotion Detection
    
    Features:
    - RoBERTa encoder with frozen lower layers
    - Multi-head attention pooling
    - Separate task-specific heads with residual connections
    - Dropout regularization throughout
    """
    def __init__(self, num_emotions=6, dropout=0.3, freeze_layers=6):
        super().__init__()
        
        # Load pre-trained RoBERTa
        self.roberta = RobertaModel.from_pretrained('roberta-base')
        hidden_size = self.roberta.config.hidden_size  # 768
        
        # Freeze lower transformer layers for efficiency
        for i, layer in enumerate(self.roberta.encoder.layer[:freeze_layers]):
            for param in layer.parameters():
                param.requires_grad = False
        
        # Multi-head attention pooling
        self.attention_pooling = MultiHeadAttentionPooling(hidden_size, num_heads=8)
        
        # Shared representation layer
        self.shared_layer = nn.Sequential(
            nn.Linear(hidden_size, hidden_size),
            nn.LayerNorm(hidden_size),
            nn.GELU(),
            nn.Dropout(dropout)
        )
        
        # Task 1: Hidden Emotion Classification Head (6 classes)
        self.emotion_head = nn.Sequential(
            nn.Linear(hidden_size, hidden_size // 2),
            nn.LayerNorm(hidden_size // 2),
            nn.GELU(),
            nn.Dropout(dropout),
            nn.Linear(hidden_size // 2, hidden_size // 4),
            nn.LayerNorm(hidden_size // 4),
            nn.GELU(),
            nn.Dropout(dropout),
            nn.Linear(hidden_size // 4, num_emotions)
        )
        
        # Task 2: Hidden Emotion Flag Head (binary)
        self.flag_head = nn.Sequential(
            nn.Linear(hidden_size, hidden_size // 2),
            nn.LayerNorm(hidden_size // 2),
            nn.GELU(),
            nn.Dropout(dropout),
            nn.Linear(hidden_size // 2, 1)
        )
        
        # Initialize weights
        self._init_weights()
    
    def _init_weights(self):
        for module in [self.shared_layer, self.emotion_head, self.flag_head]:
            for m in module.modules():
                if isinstance(m, nn.Linear):
                    nn.init.xavier_uniform_(m.weight)
                    if m.bias is not None:
                        nn.init.zeros_(m.bias)
    
    def forward(self, input_ids, attention_mask, return_attention=False):
        # Get RoBERTa outputs
        outputs = self.roberta(input_ids=input_ids, attention_mask=attention_mask)
        hidden_states = outputs.last_hidden_state  # (batch, seq_len, 768)
        
        # Multi-head attention pooling
        pooled, attn_weights = self.attention_pooling(hidden_states, attention_mask)
        
        # Shared representation with residual
        shared = self.shared_layer(pooled) + pooled
        
        # Task-specific outputs
        emotion_logits = self.emotion_head(shared)
        flag_logits = self.flag_head(shared).squeeze(-1)
        
        if return_attention:
            return emotion_logits, flag_logits, attn_weights
        return emotion_logits, flag_logits


# Initialize model
model = HiddenEmotionDetector(num_emotions=6, dropout=0.3, freeze_layers=6)
model = model.to(device)

# Count parameters
total_params = sum(p.numel() for p in model.parameters())
trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f"📊 Total parameters: {total_params:,}")
print(f"📊 Trainable parameters: {trainable_params:,}")
print(f"📊 Frozen parameters: {total_params - trainable_params:,}")

In [ ]:
# ============================================================
# TRAINING CONFIGURATION
# ============================================================

# Hyperparameters
BATCH_SIZE = 16
EPOCHS = 10
LEARNING_RATE = 2e-5
WARMUP_RATIO = 0.1
MAX_LENGTH = 128
GRADIENT_ACCUMULATION_STEPS = 2
EMOTION_LOSS_WEIGHT = 0.7  # Weight for emotion classification loss
FLAG_LOSS_WEIGHT = 0.3     # Weight for flag prediction loss

# Create datasets
train_dataset = HiddenEmotionDataset(
    train_df['text'], train_df['emotion_id'], train_df['flag'], 
    tokenizer, MAX_LENGTH
)
val_dataset = HiddenEmotionDataset(
    val_df['text'], val_df['emotion_id'], val_df['flag'], 
    tokenizer, MAX_LENGTH
)
test_dataset = HiddenEmotionDataset(
    test_df['text'], test_df['emotion_id'], test_df['flag'], 
    tokenizer, MAX_LENGTH
)

# Create dataloaders
train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True, num_workers=0)
val_loader = DataLoader(val_dataset, batch_size=BATCH_SIZE, shuffle=False, num_workers=0)
test_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False, num_workers=0)

print(f"📦 Train batches: {len(train_loader)}")
print(f"📦 Val batches: {len(val_loader)}")
print(f"📦 Test batches: {len(test_loader)}")

In [ ]:
# ============================================================
# LOSS FUNCTIONS & OPTIMIZER
# ============================================================

# Compute class weights for imbalanced data
class_weights = compute_class_weight(
    'balanced', 
    classes=np.array(list(range(6))), 
    y=train_df['emotion_id'].values
)
class_weights = torch.tensor(class_weights, dtype=torch.float).to(device)
print(f"📊 Class weights: {dict(zip(EMOTION_LABELS, class_weights.cpu().numpy().round(2)))}")

# Focal Loss for emotion classification (handles imbalance)
emotion_criterion = FocalLoss(alpha=class_weights, gamma=2.0)

# Binary Cross Entropy for flag prediction
flag_criterion = nn.BCEWithLogitsLoss()

# Optimizer with different learning rates
optimizer = torch.optim.AdamW([
    {'params': model.roberta.parameters(), 'lr': LEARNING_RATE * 0.1},  # Lower LR for pretrained
    {'params': model.attention_pooling.parameters(), 'lr': LEARNING_RATE},
    {'params': model.shared_layer.parameters(), 'lr': LEARNING_RATE},
    {'params': model.emotion_head.parameters(), 'lr': LEARNING_RATE},
    {'params': model.flag_head.parameters(), 'lr': LEARNING_RATE},
], weight_decay=0.01)

# Learning rate scheduler with warmup
total_steps = len(train_loader) * EPOCHS // GRADIENT_ACCUMULATION_STEPS
warmup_steps = int(total_steps * WARMUP_RATIO)
scheduler = get_linear_schedule_with_warmup(optimizer, warmup_steps, total_steps)

print(f"📈 Total training steps: {total_steps}")
print(f"📈 Warmup steps: {warmup_steps}")

In [ ]:
# ============================================================
# TRAINING & EVALUATION FUNCTIONS
# ============================================================

def train_epoch(model, loader, optimizer, scheduler, emotion_criterion, flag_criterion, device):
    model.train()
    total_loss = 0
    emotion_preds, emotion_labels = [], []
    flag_preds, flag_labels = [], []
    
    progress_bar = tqdm(loader, desc="Training")
    optimizer.zero_grad()
    
    for step, batch in enumerate(progress_bar):
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        emotion_label = batch['emotion_label'].to(device)
        flag_label = batch['flag_label'].to(device)
        
        # Forward pass
        emotion_logits, flag_logits = model(input_ids, attention_mask)
        
        # Compute losses
        emotion_loss = emotion_criterion(emotion_logits, emotion_label)
        flag_loss = flag_criterion(flag_logits, flag_label)
        loss = EMOTION_LOSS_WEIGHT * emotion_loss + FLAG_LOSS_WEIGHT * flag_loss
        loss = loss / GRADIENT_ACCUMULATION_STEPS
        
        # Backward pass
        loss.backward()
        
        # Gradient accumulation
        if (step + 1) % GRADIENT_ACCUMULATION_STEPS == 0:
            torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
            optimizer.step()
            scheduler.step()
            optimizer.zero_grad()
        
        total_loss += loss.item() * GRADIENT_ACCUMULATION_STEPS
        
        # Collect predictions
        emotion_preds.extend(emotion_logits.argmax(dim=1).cpu().numpy())
        emotion_labels.extend(emotion_label.cpu().numpy())
        flag_preds.extend((torch.sigmoid(flag_logits) > 0.5).cpu().numpy().astype(int))
        flag_labels.extend(flag_label.cpu().numpy().astype(int))
        
        progress_bar.set_postfix({'loss': f'{loss.item() * GRADIENT_ACCUMULATION_STEPS:.4f}'})
    
    # Compute metrics
    emotion_f1 = f1_score(emotion_labels, emotion_preds, average='weighted')
    flag_f1 = f1_score(flag_labels, flag_preds, average='binary')
    
    return total_loss / len(loader), emotion_f1, flag_f1


def evaluate(model, loader, emotion_criterion, flag_criterion, device):
    model.eval()
    total_loss = 0
    emotion_preds, emotion_labels = [], []
    flag_preds, flag_labels = [], []
    
    with torch.no_grad():
        for batch in tqdm(loader, desc="Evaluating"):
            input_ids = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            emotion_label = batch['emotion_label'].to(device)
            flag_label = batch['flag_label'].to(device)
            
            emotion_logits, flag_logits = model(input_ids, attention_mask)
            
            emotion_loss = emotion_criterion(emotion_logits, emotion_label)
            flag_loss = flag_criterion(flag_logits, flag_label)
            loss = EMOTION_LOSS_WEIGHT * emotion_loss + FLAG_LOSS_WEIGHT * flag_loss
            
            total_loss += loss.item()
            
            emotion_preds.extend(emotion_logits.argmax(dim=1).cpu().numpy())
            emotion_labels.extend(emotion_label.cpu().numpy())
            flag_preds.extend((torch.sigmoid(flag_logits) > 0.5).cpu().numpy().astype(int))
            flag_labels.extend(flag_label.cpu().numpy().astype(int))
    
    emotion_f1 = f1_score(emotion_labels, emotion_preds, average='weighted')
    flag_f1 = f1_score(flag_labels, flag_preds, average='binary')
    
    return total_loss / len(loader), emotion_f1, flag_f1, emotion_preds, emotion_labels, flag_preds, flag_labels

print("✅ Training functions defined")

In [ ]:
# ============================================================
# TRAINING LOOP
# ============================================================

print("🚀 Starting Training...")
print("=" * 60)

best_val_f1 = 0
best_model_state = None
history = {'train_loss': [], 'val_loss': [], 'train_emotion_f1': [], 'val_emotion_f1': [], 
           'train_flag_f1': [], 'val_flag_f1': []}

for epoch in range(EPOCHS):
    print(f"\n📅 Epoch {epoch + 1}/{EPOCHS}")
    print("-" * 40)
    
    # Training
    train_loss, train_emotion_f1, train_flag_f1 = train_epoch(
        model, train_loader, optimizer, scheduler, 
        emotion_criterion, flag_criterion, device
    )
    
    # Validation
    val_loss, val_emotion_f1, val_flag_f1, _, _, _, _ = evaluate(
        model, val_loader, emotion_criterion, flag_criterion, device
    )
    
    # Store history
    history['train_loss'].append(train_loss)
    history['val_loss'].append(val_loss)
    history['train_emotion_f1'].append(train_emotion_f1)
    history['val_emotion_f1'].append(val_emotion_f1)
    history['train_flag_f1'].append(train_flag_f1)
    history['val_flag_f1'].append(val_flag_f1)
    
    print(f"📉 Train Loss: {train_loss:.4f} | Val Loss: {val_loss:.4f}")
    print(f"🎯 Train Emotion F1: {train_emotion_f1:.4f} | Val Emotion F1: {val_emotion_f1:.4f}")
    print(f"🏳️ Train Flag F1: {train_flag_f1:.4f} | Val Flag F1: {val_flag_f1:.4f}")
    
    # Save best model
    combined_f1 = 0.7 * val_emotion_f1 + 0.3 * val_flag_f1
    if combined_f1 > best_val_f1:
        best_val_f1 = combined_f1
        best_model_state = model.state_dict().copy()
        print(f"✅ New best model saved! Combined F1: {combined_f1:.4f}")

# Load best model
model.load_state_dict(best_model_state)
print(f"\n🏆 Training complete! Best Combined F1: {best_val_f1:.4f}")

In [ ]:
# ============================================================
# TRAINING VISUALIZATION
# ============================================================

fig, axes = plt.subplots(1, 3, figsize=(15, 4))

# Loss curves
ax1 = axes[0]
ax1.plot(history['train_loss'], label='Train Loss', marker='o')
ax1.plot(history['val_loss'], label='Val Loss', marker='s')
ax1.set_xlabel('Epoch')
ax1.set_ylabel('Loss')
ax1.set_title('Training & Validation Loss')
ax1.legend()
ax1.grid(True, alpha=0.3)

# Emotion F1 curves
ax2 = axes[1]
ax2.plot(history['train_emotion_f1'], label='Train F1', marker='o')
ax2.plot(history['val_emotion_f1'], label='Val F1', marker='s')
ax2.set_xlabel('Epoch')
ax2.set_ylabel('F1 Score')
ax2.set_title('Emotion Classification F1')
ax2.legend()
ax2.grid(True, alpha=0.3)

# Flag F1 curves
ax3 = axes[2]
ax3.plot(history['train_flag_f1'], label='Train F1', marker='o')
ax3.plot(history['val_flag_f1'], label='Val F1', marker='s')
ax3.set_xlabel('Epoch')
ax3.set_ylabel('F1 Score')
ax3.set_title('Hidden Emotion Flag F1')
ax3.legend()
ax3.grid(True, alpha=0.3)

plt.tight_layout()
plt.show()

In [ ]:
# ============================================================
# FINAL EVALUATION ON TEST SET
# ============================================================

print("🧪 Evaluating on Test Set...")
print("=" * 60)

test_loss, test_emotion_f1, test_flag_f1, emotion_preds, emotion_labels, flag_preds, flag_labels = evaluate(
    model, test_loader, emotion_criterion, flag_criterion, device
)

print(f"\n📊 TEST RESULTS")
print("=" * 60)
print(f"Test Loss: {test_loss:.4f}")
print(f"Emotion Classification F1 (weighted): {test_emotion_f1:.4f}")
print(f"Hidden Emotion Flag F1: {test_flag_f1:.4f}")

# Detailed classification reports
print("\n" + "=" * 60)
print("EMOTION CLASSIFICATION REPORT")
print("=" * 60)
print(classification_report(emotion_labels, emotion_preds, target_names=EMOTION_LABELS, digits=4))

print("\n" + "=" * 60)
print("HIDDEN EMOTION FLAG REPORT")
print("=" * 60)
print(classification_report(flag_labels, flag_preds, target_names=['No Hidden', 'Has Hidden'], digits=4))

In [ ]:
# ============================================================
# CONFUSION MATRIX VISUALIZATION
# ============================================================

import seaborn as sns

fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# Emotion confusion matrix
ax1 = axes[0]
cm_emotion = confusion_matrix(emotion_labels, emotion_preds)
sns.heatmap(cm_emotion, annot=True, fmt='d', cmap='Blues', 
            xticklabels=EMOTION_LABELS, yticklabels=EMOTION_LABELS, ax=ax1)
ax1.set_xlabel('Predicted')
ax1.set_ylabel('Actual')
ax1.set_title('Hidden Emotion Classification\nConfusion Matrix')

# Flag confusion matrix
ax2 = axes[1]
cm_flag = confusion_matrix(flag_labels, flag_preds)
sns.heatmap(cm_flag, annot=True, fmt='d', cmap='Greens',
            xticklabels=['No Hidden', 'Has Hidden'], 
            yticklabels=['No Hidden', 'Has Hidden'], ax=ax2)
ax2.set_xlabel('Predicted')
ax2.set_ylabel('Actual')
ax2.set_title('Hidden Emotion Flag\nConfusion Matrix')

plt.tight_layout()
plt.show()

In [ ]:
# ============================================================
# INFERENCE FUNCTION WITH ATTENTION VISUALIZATION
# ============================================================

def predict_hidden_emotion(text, model, tokenizer, device, show_attention=False):
    """
    Predict hidden emotion for a single text input
    
    Returns:
        dict with predictions and confidence scores
    """
    model.eval()
    
    # Tokenize
    encoding = tokenizer(
        text,
        truncation=True,
        max_length=128,
        padding='max_length',
        return_tensors='pt'
    )
    
    input_ids = encoding['input_ids'].to(device)
    attention_mask = encoding['attention_mask'].to(device)
    
    with torch.no_grad():
        emotion_logits, flag_logits, attn_weights = model(
            input_ids, attention_mask, return_attention=True
        )
        
        # Emotion prediction
        emotion_probs = F.softmax(emotion_logits, dim=1).squeeze()
        emotion_pred = emotion_probs.argmax().item()
        emotion_confidence = emotion_probs[emotion_pred].item()
        
        # Flag prediction
        flag_prob = torch.sigmoid(flag_logits).item()
        has_hidden_emotion = flag_prob > 0.5
    
    result = {
        'text': text,
        'hidden_emotion': id2emotion[emotion_pred],
        'emotion_confidence': emotion_confidence,
        'emotion_probabilities': {id2emotion[i]: p.item() for i, p in enumerate(emotion_probs)},
        'has_hidden_emotion': has_hidden_emotion,
        'hidden_emotion_probability': flag_prob
    }
    
    if show_attention:
        # Get tokens for attention visualization
        tokens = tokenizer.convert_ids_to_tokens(input_ids[0])
        valid_len = attention_mask[0].sum().item()
        result['attention_weights'] = attn_weights[0, 0, :valid_len].cpu().numpy()
        result['tokens'] = tokens[:valid_len]
    
    return result


# Test inference
print("🔮 Testing Inference Function")
print("=" * 60)

test_texts = [
    "I love how everyone ignores my messages 😊",
    "Best day ever! Finally got promoted! 🎉",
    "Sure, I'm totally fine with working overtime again 😊",
    "Can't believe they canceled my favorite show 😭"
]

for text in test_texts:
    result = predict_hidden_emotion(text, model, tokenizer, device)
    print(f"\n📝 Text: {text}")
    print(f"   🎭 Hidden Emotion: {result['hidden_emotion']} ({result['emotion_confidence']:.2%})")
    print(f"   🏳️ Has Hidden Emotion: {result['has_hidden_emotion']} ({result['hidden_emotion_probability']:.2%})")
    print(f"   📊 All probs: {', '.join([f'{e}: {p:.1%}' for e, p in result['emotion_probabilities'].items()])}")

In [ ]:
# ============================================================
# SAVE MODEL
# ============================================================

import json

# Save model weights
model_save_path = "/Users/Rivin/Desktop/untitled folder/Final/hidden_emotion_model.pt"
torch.save({
    'model_state_dict': model.state_dict(),
    'emotion2id': emotion2id,
    'id2emotion': id2emotion,
    'config': {
        'num_emotions': 6,
        'dropout': 0.3,
        'freeze_layers': 6,
        'max_length': MAX_LENGTH
    }
}, model_save_path)

# Save training history
history_path = "/Users/Rivin/Desktop/untitled folder/Final/training_history.json"
with open(history_path, 'w') as f:
    json.dump(history, f)

print(f"✅ Model saved to: {model_save_path}")
print(f"✅ Training history saved to: {history_path}")

# Model summary
print("\n" + "=" * 60)
print("📊 MODEL SUMMARY")
print("=" * 60)
print(f"Architecture: RoBERTa-base + Multi-Head Attention Pooling")
print(f"Tasks: Hidden Emotion (6-class) + Flag (binary)")
print(f"Total Parameters: {total_params:,}")
print(f"Trainable Parameters: {trainable_params:,}")
print(f"Best Validation F1: {best_val_f1:.4f}")
print(f"Test Emotion F1: {test_emotion_f1:.4f}")
print(f"Test Flag F1: {test_flag_f1:.4f}")

## 🏆 Model Architecture Summary

### HiddenEmotionDetector
```
┌─────────────────────────────────────────────────────────────┐
│                    INPUT TEXT                                │
│              "I love how everyone ignores me 😊"             │
└─────────────────────────────────────────────────────────────┘
                            │
                            ▼
┌─────────────────────────────────────────────────────────────┐
│              RoBERTa Tokenizer (max_len=128)                │
└─────────────────────────────────────────────────────────────┘
                            │
                            ▼
┌─────────────────────────────────────────────────────────────┐
│                   RoBERTa Encoder                           │
│        (12 layers, 768 hidden, 6 frozen layers)             │
└─────────────────────────────────────────────────────────────┘
                            │
                            ▼
┌─────────────────────────────────────────────────────────────┐
│           Multi-Head Attention Pooling (8 heads)            │
│              (Learnable query + attention)                  │
└─────────────────────────────────────────────────────────────┘
                            │
                            ▼
┌─────────────────────────────────────────────────────────────┐
│                 Shared Layer + Residual                     │
│            (Linear → LayerNorm → GELU → Dropout)            │
└─────────────────────────────────────────────────────────────┘
                     │                │
                     ▼                ▼
        ┌────────────────┐  ┌────────────────┐
        │  Emotion Head  │  │   Flag Head    │
        │ (768→384→192→6)│  │  (768→384→1)   │
        └────────────────┘  └────────────────┘
                │                    │
                ▼                    ▼
        ┌────────────────┐  ┌────────────────┐
        │ 6-class output │  │ Binary output  │
        │  (softmax)     │  │  (sigmoid)     │
        └────────────────┘  └────────────────┘
```

### Key Features:
1. **Focal Loss** - Handles class imbalance (γ=2.0)
2. **Class Weights** - Computed from training distribution
3. **Layer Freezing** - First 6 transformer layers frozen
4. **Gradient Accumulation** - Effective batch size = 32
5. **Warmup Scheduler** - Linear warmup + decay
6. **Multi-Task Learning** - 70% emotion + 30% flag loss